ticketutil/servicenow.py

```diff
@@ -0,0 +1,323 @@
+api PATCH method would enable editing comment
+api DELETE method - maybe, not that necessary
+"""
+
+
+import logging
+import os
+
+from . import ticket
+
+__author__ = 'dranck, rnester, kshirsal, pzubaty'
+
+DEBUG = os.environ.get('TICKETUTIL_DEBUG', 'False')
+
+if DEBUG == 'True':
+    logging.basicConfig(level=logging.DEBUG)
+else:
+    logging.basicConfig(level=logging.INFO)
+
+
+class ServiceNowTicket(ticket.Ticket):
+    """
+    ServiceNow Ticket object. Contains ServiceNow specific methods for working
+    with tickets.
+    """
+
+    def __init__(self, url, project, auth=None, ticket_id=None):
+        """
+        :param url: ServiceNow service url
+        :param project: ServiceNow table or project
+        :param auth: (<username>, <password>) for HTTP Basic Authentication
+        :param ticked_id: ticked number, eg. 'PNT1234567'
+        """
+        self.ticketing_tool = 'ServiceNow'
+        self.sys_id = None
+
+        # The auth param should be of the form (<username>, <password>) for
+        # HTTP Basic authentication.
+        self.url = url
+        self.auth = auth
+        self.table = project
+        self.rest_url = '{0}/api/now/v1/table/{1}'.format(self.url, self.table)
+        self.auth_url = self.rest_url
+
+        # parent creates request session, ticket_id beforehand causes failure
+        super(ServiceNowTicket, self).__init__(project, ticket_id=None)
+        if ticket_id:
+            self.ticket_id = ticket_id
+            self._get_ticket_content()
+            self.sys_id = self.ticket_content['sys_id']
+            super(ServiceNowTicket, self).set_ticket_id(ticket_id)
+
+    def _get_ticket_content(self):
+        """
+        Get sys_id from ticket_id
+
+        :return: sys_id - hashcode identifier in the ticket URL
+        """
+        try:
+            self.s.headers.update({'Content-Type': 'application/json'})
+            url = self.rest_url + '?sysparm_query=GOTOnumber%3D'
+            url += self.ticket_id
+            r = self.s.get(url)
+            r.raise_for_status()
+
+            logging.debug("Get ticket content: Status Code: {0}".format(r.status_code))
+            ticket_content = r.json()
+            print(ticket_content)
+            self.ticket_content = ticket_content['result'][0]
+        except requests.RequestException as e:
+            logging.error("Error while getting ticket content")
+            logging.error(e.args[0])
+
+    def _generate_ticket_url(self):
+        """
+        Generates the ticket URL out of the url, project, and ticket_id.
+
+        :return: ticket_url: The URL of the ticket.
+        """
+        ticket_url = None
+
+        # If we are receiving a ticket_id, we have sys_id
+        if self.sys_id:
+            ticket_url = '{0}/{1}.do?sys_id={2}'.format(self.url, self.table,
+                                                        self.sys_id)
+        return ticket_url
+
+    def create(self, subject, description, category, item, **kwargs):
```

Comment:
Category and item are mandatory for standard ServiceNow WebUI, but optional for PNT DevOps One.